File: PQDIF.Validator/DataSource/CustomSourceInfo.cs
/*
**  Class:          PQDIF.Validator.CustomSourceInfo
**  Description:	One of these collections per channel
**
** --------------------------------------------------------------------------
**
** Copyright 2022 PQDIF Authors
**
** Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at
**
**      http://www.apache.org/licenses/LICENSE-2.0
**
** Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
** either express or implied. See the License for the specific language governing permissions and limitations under the License.
**
** See the LICENSE file distributed with this work for copyright and licensing information, the AUTHORS file for a list of
** copyright holders, and the CONTRIBUTORS file for the list of contributors.
**
** SPDX - License - Identifier: Apache - 2.0
*/


// Contains fundamental classes and base classes that define commonly-used value and reference data types, events and event handlers, interfaces, attributes, and processing exceptions.
using System;

// Defines IEEE PQDIF data types including PQDIF timestamps
using PQDIFNet;

// Defines IEEE PQDIF IDs and data types 
using static PQDIFNet.Constants.Logical;

// For logging errors, warnings, other messages.
using PQDIF.Log;


namespace PQDIF.Validator
{

    /// <summary>
    /// This collection can include the standard name, address and telephone number tags. These apply to the vendor as well as tags about the instrument itself
    /// </summary>
    public class CustomSourceInfo
    {
        /// <summary>
        /// The type of instrument
        /// </summary>
        public Guid InstrumentTypeID = ID_INSTR_TYPE_PQM;

        /// <summary>
        /// Arbitrary string
        /// </summary>
        public string InstrumentModelName = "";

        /// <summary>
        /// Arbitrary string
        /// </summary>
        public string InstrumentModelNumber = "";

        /// <summary>
        /// Optional Contact Name
        /// </summary>
        public string Name = "";

        /// <summary>
        /// Optional Line 1 of a Mailing Address
        /// </summary>
        public string Address1 = "";

        /// <summary>
        /// Optional Line 2 of a Mailing Address
        /// </summary>
        public string Address2 = "";

        /// <summary>
        /// Optional City of a Mailing Address
        /// </summary>
        public string City = "";

        /// <summary>
        /// Optional State or Province
        /// </summary>
        public string State = "";

        /// <summary>
        /// Optional Postal Code
        /// </summary>
        public string PostalCode = "";

        /// <summary>
        /// Optional Country of a Mailing Address
        /// </summary>
        public string Country = "";

        /// <summary>
        /// Optional Telephone Number
        /// </summary>
        public string PhoneVoice = "";

        /// <summary>
        /// Optional Fax Number
        /// </summary>
        public string PhoneFAX = "";

        /// <summary>
        /// Optional Email Address
        /// </summary>
        public string EMail = "";

        /// <summary>
        /// IEEE Std 1159.3 Compliance Logger for this <see cref="CustomSourceInfo"/>
        /// </summary>
        /// <remarks>Not an IEEE 1159.3 Tag</remarks>
        public Logger LoggerCompliance = new Logger();

        /// <summary>
        /// Application Logger for this <see cref="CustomSourceInfo"/>
        /// </summary>
        /// <remarks>Not an IEEE 1159.3 Tag</remarks>
        public Logger LoggerApplication = new Logger();

        /// <summary>
        /// Initializes a new instance of the <see cref="CustomSourceInfo"/> class.
        /// </summary>
        public CustomSourceInfo()
        {
            // Default Constructor
        }

        /// <summary>
        /// Initializes a new instance of the <see cref="CustomSourceInfo"/> class.
        /// </summary>
        /// <param name="PqdifFile">PQDIF file from which to read a container record</param>
        /// <param name="PointerCollection">Pointer to the Collection that Contains a Data Source Record in <paramref name="PqdifFile"/></param>
        /// <param name="LogLevelApplication">Specifies the Log Level for Application Errors</param>
        /// <param name="LogLevelCompliance">Specifies the Log Level for Compliance Errors</param>
        /// <param name="RecordOK"><c>true</c> if the record was read OK</param>
        public CustomSourceInfo(CPQDIFNet PqdifFile, ref IntPtr PointerCollection, LogLevels LogLevelApplication, LogLevels LogLevelCompliance, out bool RecordOK)
        {
            // Initialize the return value
            RecordOK = true;

            // Set up an error handler
            try
            {
                // Set the logger level
                LoggerApplication.LogLevel = LogLevelApplication;
                LoggerCompliance.LogLevel = LogLevelCompliance;

                // Determine the number of tags associated with the current record
                int CollectionCount = 0;
                bool Result = PqdifFile.CollectionGetCount(PointerCollection, ref CollectionCount);
                if (!Result)
                {
                    LoggerApplication.Log("CollectionGetCount Error", LogLevels.Error);
                    RecordOK = false;
                    return;
                }

                // For each tag in the CustomSourceInfo collectione...
                for (int CollectionIndex = 0; CollectionIndex < CollectionCount; CollectionIndex++)
                {
                    // Get the pointer to the current tag
                    IntPtr PointerCollectionTag = new IntPtr();
                    Result = PqdifFile.CollectionGetEntry(PointerCollection, CollectionIndex, ref PointerCollectionTag);
                    if (!Result)
                    {
                        LoggerApplication.Log("CollectionGetEntry", LogLevels.Error);
                        RecordOK = false;
                        return;
                    }

                    // Get the tag ID and tag name
                    Guid TagID = new Guid();
                    string TagName = "";
                    Result = PqdifFile.ElementGetTag(PointerCollectionTag, ref TagID, ref TagName);
                    if (!Result)
                    {
                        LoggerApplication.Log("ElementGetTag", LogLevels.Error);
                        RecordOK = false;
                        return;
                    }

                    // Initialize a flag that tells us that the current TagID was not matched to a known TagID.
                    bool TagMatched = false;

                    // If the current TagID has not been matched, then try to match it as all other known tags of the Container Collection.
                    if (!TagMatched) TagMatched = PQDIFNetWrapper.ElementGetScalarData(TagID, tagInstrumentTypeID, PqdifFile, PointerCollectionTag, LoggerApplication, LoggerCompliance, "", ref InstrumentTypeID);
                    if (!TagMatched) TagMatched = PQDIFNetWrapper.ElementGetString(TagID, tagInstrumentModelName, PqdifFile, PointerCollectionTag, LoggerApplication, LoggerCompliance, "", ref InstrumentModelName);
                    if (!TagMatched) TagMatched = PQDIFNetWrapper.ElementGetString(TagID, tagInstrumentModelNumber, PqdifFile, PointerCollectionTag, LoggerApplication, LoggerCompliance, "", ref InstrumentModelNumber);
                    if (!TagMatched) TagMatched = PQDIFNetWrapper.ElementGetString(TagID, tagName, PqdifFile, PointerCollectionTag, LoggerApplication, LoggerCompliance, "", ref Name);
                    if (!TagMatched) TagMatched = PQDIFNetWrapper.ElementGetString(TagID, tagAddress1, PqdifFile, PointerCollectionTag, LoggerApplication, LoggerCompliance, "", ref Address1);
                    if (!TagMatched) TagMatched = PQDIFNetWrapper.ElementGetString(TagID, tagAddress2, PqdifFile, PointerCollectionTag, LoggerApplication, LoggerCompliance, "", ref Address2);
                    if (!TagMatched) TagMatched = PQDIFNetWrapper.ElementGetString(TagID, tagCity, PqdifFile, PointerCollectionTag, LoggerApplication, LoggerCompliance, "", ref City);
                    if (!TagMatched) TagMatched = PQDIFNetWrapper.ElementGetString(TagID, tagState, PqdifFile, PointerCollectionTag, LoggerApplication, LoggerCompliance, "", ref State);
                    if (!TagMatched) TagMatched = PQDIFNetWrapper.ElementGetString(TagID, tagPostalCode, PqdifFile, PointerCollectionTag, LoggerApplication, LoggerCompliance, "", ref PostalCode);
                    if (!TagMatched) TagMatched = PQDIFNetWrapper.ElementGetString(TagID, tagCountry, PqdifFile, PointerCollectionTag, LoggerApplication, LoggerCompliance, "", ref Country);
                    if (!TagMatched) TagMatched = PQDIFNetWrapper.ElementGetString(TagID, tagPhoneVoice, PqdifFile, PointerCollectionTag, LoggerApplication, LoggerCompliance, "", ref PhoneVoice);
                    if (!TagMatched) TagMatched = PQDIFNetWrapper.ElementGetString(TagID, tagPhoneFAX, PqdifFile, PointerCollectionTag, LoggerApplication, LoggerCompliance, "", ref PhoneFAX);
                    if (!TagMatched) TagMatched = PQDIFNetWrapper.ElementGetString(TagID, tagEMail, PqdifFile, PointerCollectionTag, LoggerApplication, LoggerCompliance, "", ref EMail);

                    // If still not matched, try to match the current tag to tagBlank, which can be used to leave a space in a collection.
                    if (!TagMatched)
                        TagMatched = TagID == tagBlank;

                    // If the TagID has not been matched, then log an information message about this unknown tag.
                    if (!TagMatched)
                        LoggerCompliance.Log(": Unknown or Unexpected Tag: " + PQDIFNetWrapper.GetName(TagID), LogLevels.Info);

                } // For each tag in the CustomSourceInfo collection

            } // try

            // Handle exceptions here.
            catch (Exception ex)
            {
                LoggerApplication.Log(ex.Message, LogLevels.Error);
            }

        } // CustomSourceInfo


        /// <summary>
        /// Returns the <see cref="CustomSourceInfo"/> Collection Formated in XML
        /// </summary>
        /// <param name="ApplicationLog">If set to <c>true</c>, application log messages will be returned</param>
        /// <param name="ComplianceLog">If set to <c>true</c>, IEEE Std 1159.3 compliance log messages will be returned</param>
        /// <value>Custom Source Info Collection Formated in XML</value>
        public string GetXML(bool ApplicationLog, bool ComplianceLog)
        {
            // Initialize a string builder
            System.Text.StringBuilder StringBuilder = new System.Text.StringBuilder();
            StringBuilder.Append("<tagCustomSourceInfo>\r\n");

            // If we should return application logs...
            if (ApplicationLog)
            {
                StringBuilder.Append("<ApplicationLog>\r\n");
                StringBuilder.Append(LoggerApplication.GetLogXML());
                StringBuilder.Append("</ApplicationLog>\r\n");
            }

            // If we should return compliance logs...
            if (ComplianceLog)
            {
                StringBuilder.Append("<ComplianceLog>\r\n");
                StringBuilder.Append(LoggerCompliance.GetLogXML());
                StringBuilder.Append("</ComplianceLog>\r\n");
            }

            // Export the tags of the Custom Source Info collection as XML elements.
            StringBuilder.Append(Utilities.XML.GetElement(tagInstrumentTypeID, InstrumentTypeID) + "\r\n");
            StringBuilder.Append(Utilities.XML.GetElement(tagInstrumentModelName, InstrumentModelName) + "\r\n");
            StringBuilder.Append(Utilities.XML.GetElement(tagInstrumentModelNumber, InstrumentModelNumber) + "\r\n");

            StringBuilder.Append(Utilities.XML.GetElement(tagName, Name) + "\r\n");
            StringBuilder.Append(Utilities.XML.GetElement(tagAddress1, Address1) + "\r\n");
            StringBuilder.Append(Utilities.XML.GetElement(tagAddress2, Address2) + "\r\n");
            StringBuilder.Append(Utilities.XML.GetElement(tagCity, City) + "\r\n");
            StringBuilder.Append(Utilities.XML.GetElement(tagState, State) + "\r\n");
            StringBuilder.Append(Utilities.XML.GetElement(tagPostalCode, PostalCode) + "\r\n");
            StringBuilder.Append(Utilities.XML.GetElement(tagCountry, Country) + "\r\n");
            StringBuilder.Append(Utilities.XML.GetElement(tagPhoneVoice, PhoneVoice) + "\r\n");
            StringBuilder.Append(Utilities.XML.GetElement(tagPhoneFAX, PhoneFAX) + "\r\n");
            StringBuilder.Append(Utilities.XML.GetElement(tagEMail, EMail) + "\r\n");
                        
            // Finish the Custom Source Info record.
            StringBuilder.Append("</tagCustomSourceInfo>\r\n");

            // Return the XML version of the Custom Source Info
            return StringBuilder.ToString();

        } // GetXML

    } // class CustomSourceInfo

} // namespace PQDIF.Validator

File: src/validator/customSourceInfo.js
'use strict';

const { PqdifInfo, guidEquals } = require('../info/info');
const { Logger } = require('../logger');
const { LogLevels } = require('../logger');
const tagGuids = require('../constants/tagGuids');
const { xmlGetElement } = require('./utilities');

/**
 * Custom source info collection - ported from CustomSourceInfo.cs.
 * Contains instrument type, model, contact info, etc.
 */
class CustomSourceInfo {
  constructor() {
    /** @type {Buffer} Instrument type GUID */
    this.instrumentTypeID = null;

    /** @type {string} Instrument model name */
    this.instrumentModelName = '';

    /** @type {string} Instrument model number */
    this.instrumentModelNumber = '';

    /** @type {string} Contact name */
    this.name = '';

    /** @type {string} Address line 1 */
    this.address1 = '';

    /** @type {string} Address line 2 */
    this.address2 = '';

    /** @type {string} City */
    this.city = '';

    /** @type {string} State or province */
    this.state = '';

    /** @type {string} Postal code */
    this.postalCode = '';

    /** @type {string} Country */
    this.country = '';

    /** @type {string} Phone number */
    this.phoneVoice = '';

    /** @type {string} Fax number */
    this.phoneFAX = '';

    /** @type {string} Email address */
    this.email = '';

    /** @type {Logger} Application logger */
    this.loggerApplication = new Logger();

    /** @type {Logger} Compliance logger */
    this.loggerCompliance = new Logger();
  }

  /**
   * Create from a Collection element.
   * @param {import('../core/collection').Collection} coll - The collection element
   */
  static fromCollection(coll) {
    const info = new CustomSourceInfo();
    if (!coll) return info;

    const getCount = coll.getCount ? coll.getCount() : 0;
    for (let i = 0; i < getCount; i++) {
      const el = coll.getElement(i);
      if (!el || !el.getTag()) continue;

      const tag = el.getTag();
      try {
        if (guidEquals(tag, tagGuids.tagInstrumentTypeID)) {
          const r = el.getValueGUID ? el.getValueGUID() : null;
          if (r && r.status) info.instrumentTypeID = r.value;
        } else if (guidEquals(tag, tagGuids.tagInstrumentModelName)) {
          info.instrumentModelName = _getStringFromElement(el);
        } else if (guidEquals(tag, tagGuids.tagInstrumentModelNumber)) {
          info.instrumentModelNumber = _getStringFromElement(el);
        } else if (guidEquals(tag, tagGuids.tagName)) {
          info.name = _getStringFromElement(el);
        } else if (guidEquals(tag, tagGuids.tagAddress1)) {
          info.address1 = _getStringFromElement(el);
        } else if (guidEquals(tag, tagGuids.tagAddress2)) {
          info.address2 = _getStringFromElement(el);
        } else if (guidEquals(tag, tagGuids.tagCity)) {
          info.city = _getStringFromElement(el);
        } else if (guidEquals(tag, tagGuids.tagState)) {
          info.state = _getStringFromElement(el);
        } else if (guidEquals(tag, tagGuids.tagPostalCode)) {
          info.postalCode = _getStringFromElement(el);
        } else if (guidEquals(tag, tagGuids.tagCountry)) {
          info.country = _getStringFromElement(el);
        } else if (guidEquals(tag, tagGuids.tagPhoneVoice)) {
          info.phoneVoice = _getStringFromElement(el);
        } else if (guidEquals(tag, tagGuids.tagPhoneFAX)) {
          info.phoneFAX = _getStringFromElement(el);
        } else if (guidEquals(tag, tagGuids.tagEMail)) {
          info.email = _getStringFromElement(el);
        } else if (guidEquals(tag, tagGuids.tagBlank)) {
          // blank tag - ignore
        } else {
          info.loggerCompliance.log('Unknown or Unexpected Tag: ' + (PqdifInfo.getTagName(tag) || tag.toString('hex')), LogLevels.Info);
        }
      } catch (ex) {
        info.loggerApplication.log('Error reading tag: ' + ex.message, LogLevels.Error);
      }
    }

    return info;
  }

  /**
   * Get XML representation.
   * @param {boolean} applicationLog
   * @param {boolean} complianceLog
   * @returns {string}
   */
  getXML(applicationLog, complianceLog) {
    let xml = '<tagCustomSourceInfo>\r\n';
    if (applicationLog) {
      xml += '<ApplicationLog>\r\n' + this.loggerApplication.getLogXML() + '</ApplicationLog>\r\n';
    }
    if (complianceLog) {
      xml += '<ComplianceLog>\r\n' + this.loggerCompliance.getLogXML() + '</ComplianceLog>\r\n';
    }
    if (this.instrumentTypeID) xml += xmlGetElement(tagGuids.tagInstrumentTypeID, this.instrumentTypeID) + '\r\n';
    xml += xmlGetElement(tagGuids.tagInstrumentModelName, this.instrumentModelName) + '\r\n';
    xml += xmlGetElement(tagGuids.tagInstrumentModelNumber, this.instrumentModelNumber) + '\r\n';
    xml += xmlGetElement(tagGuids.tagName, this.name) + '\r\n';
    xml += xmlGetElement(tagGuids.tagAddress1, this.address1) + '\r\n';
    xml += xmlGetElement(tagGuids.tagAddress2, this.address2) + '\r\n';
    xml += xmlGetElement(tagGuids.tagCity, this.city) + '\r\n';
    xml += xmlGetElement(tagGuids.tagState, this.state) + '\r\n';
    xml += xmlGetElement(tagGuids.tagPostalCode, this.postalCode) + '\r\n';
    xml += xmlGetElement(tagGuids.tagCountry, this.country) + '\r\n';
    xml += xmlGetElement(tagGuids.tagPhoneVoice, this.phoneVoice) + '\r\n';
    xml += xmlGetElement(tagGuids.tagPhoneFAX, this.phoneFAX) + '\r\n';
    xml += xmlGetElement(tagGuids.tagEMail, this.email) + '\r\n';
    xml += '</tagCustomSourceInfo>\r\n';
    return xml;
  }

  /**
   * Get JSON representation.
   * @returns {Object}
   */
  getJSON() {
    return {
      instrumentTypeID: this.instrumentTypeID ? this.instrumentTypeID.toString('hex') : null,
      instrumentModelName: this.instrumentModelName || '',
      instrumentModelNumber: this.instrumentModelNumber || '',
      contact: {
        name: this.name || '',
        address1: this.address1 || '',
        address2: this.address2 || '',
        city: this.city || '',
        state: this.state || '',
        postalCode: this.postalCode || '',
        country: this.country || '',
        phoneVoice: this.phoneVoice || '',
        phoneFAX: this.phoneFAX || '',
        email: this.email || ''
      }
    };
  }
}

/**
 * Extract string from a Vector element.
 * @param {import('../core/vector').Vector} el
 * @returns {string}
 */
function _getStringFromElement(el) {
  if (el && typeof el.getValues === 'function') {
    const out = { value: '' };
    el.getValues(out);
    return out.value;
  }
  if (el && typeof el.getString === 'function') {
    return el.getString();
  }
  return '';
}

module.exports = CustomSourceInfo;
